src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+                "cannot commit a transaction in a read-only catalog".to_string(),
+            )
+            .into());
+        }
+
+        let current_upper = self.upper.clone();
+        let next_upper = current_upper + 1;
+
+        let updates = StateUpdate::from_txn_batch(txn_batch, current_upper);
+        self.apply_updates(updates.clone().into_iter())?;
+
+        if matches!(self.mode, Mode::Writable) {
+            let mut batch_builder = self
+                .persist_handle
+                .write_handle
+                .builder(Antichain::from_elem(current_upper));
+
+            for StateUpdate { kind, ts, diff } in updates {
+                batch_builder
+                    .add(&kind, &(), &ts, &diff)
+                    .await
+                    .expect("invalid usage");
+            }
+            // If we haven't fenced the previous catalog state, do that now.
+            if !self.fenced {
+                batch_builder
+                    .add(&StateUpdateKind::Epoch(self.epoch), &(), &current_upper, &1)
+                    .await
+                    .expect("invalid usage");
+            }
+            let mut batch = batch_builder
+                .finish(Antichain::from_elem(next_upper))
+                .await
+                .expect("invalid usage");
+            match self
+                .persist_handle
+                .write_handle
+                .compare_and_append_batch(
+                    &mut [&mut batch],
+                    Antichain::from_elem(current_upper),
+                    Antichain::from_elem(next_upper),
+                )
+                .await
+                .expect("invalid usage")
+            {
+                Ok(()) => {
+                    self.persist_handle
+                        .read_handle
+                        .downgrade_since(&Antichain::from_elem(current_upper))
+                        .await;
+                    self.upper = next_upper;
+                }
+                Err(upper_mismatch) => {
+                    return Err(DurableCatalogError::Fence(format!(
+                        "current catalog upper {:?} fenced by new catalog upper {:?}",
+                        upper_mismatch.expected, upper_mismatch.current
+                    ))
+                    .into())
+                }
+            }
+        }
+
+        self.fenced = true;
+        Ok(())
+    }
+
+    async fn confirm_leadership(&mut self) -> Result<(), CatalogError> {
+        let upper = self.persist_handle.current_upper().await;
+        if upper == self.upper {
+            Ok(())
+        } else {
+            Err(DurableCatalogError::Fence(format!(
+                "current catalog upper {:?} fenced by new catalog upper {:?}",
+                self.upper, upper
+            ))
+            .into())
+        }
+    }
+
+    async fn set_connect_timeout(&mut self, _connect_timeout: Duration) {
+        // TODO(jkosh44) Is there an equivalent in persist?
```

Comment:
I believe it's done internally. Do we need to expose control of it for some reason?

Reply:
No, I updated this comment. Both Persist and Stash use the `crdb_connect_timeout` system variable to control CRDB timeouts. So this should be a no-op.